src/main/java/org/opensearch/security/dlic/rest/api/AccountApiAction.java

```diff
@@ -183,47 +237,106 @@ protected void handleGet(RestChannel channel, RestRequest request, Client client
+            notFound(channel, "Could not find user.");
             return;
         }
 
-        final SecurityJsonNode securityJsonNode = new SecurityJsonNode(content);
-        final String currentPassword = content.get("current_password").asText();
-        final Hashed internalUserEntry = (Hashed) internalUser.getCEntry(username);
-        final String currentHash = internalUserEntry.getHash();
-
-        if (currentHash == null || !OpenBSDBCrypt.checkPassword(currentHash, currentPassword.toCharArray())) {
-            badRequestResponse(channel, "Could not validate your current password.");
+        if (!isWriteable(channel, internalUser, user.getName())) {
             return;
         }
 
-        // if password is set, it takes precedence over hash
-        final String password = securityJsonNode.get("password").asString();
-        final String hash;
-        if (Strings.isNullOrEmpty(password)) {
-            hash = securityJsonNode.get("hash").asString();
-        } else {
-            hash = hash(password.toCharArray());
-        }
-        if (Strings.isNullOrEmpty(hash)) {
-            badRequestResponse(channel, "Both provided password and hash cannot be null/empty.");
+        // request must have 'saved_tenant' AND/OR ('current_password' AND ('password' OR 'hash'))
+        final boolean invalidSavedTenantBody = content.get("saved_tenant") == null;
+        final boolean invalidPasswordBody = content.get("current_password") == null || (content.get("password") == null && content.get("hash") == null);
+        // bad case: no valid saved_tenant body nor password body
+        if (invalidSavedTenantBody && invalidPasswordBody){
+            badRequestResponse(channel, "Invalid request. Needs a saved_tenant AND/OR (saved_password AND (password OR hash)");
             return;
         }
 
-        internalUserEntry.setHash(hash);
+        // process PUT for 'saved_tenant'
+        if (content.get("saved_tenant") != null){
+            final String newSavedTenant = content.get("saved_tenant").asText();
+            // invalid action; user is not InternalUserV7
+            if (!(internalUser.getCEntry(user.getName()) instanceof InternalUserV7)){
+                badRequestResponse(channel, "Cannot modify saved_tenant for non-InternalUserV7.");
+                return;
+            }
+            InternalUserV7 targetUser = (InternalUserV7) internalUser.getCEntry(user.getName());
+            // each user always has access to the global tenant and their own private tenant
+            if (!(newSavedTenant.equals(DEFAULT_TENANT) || newSavedTenant.equals(PRIVATE_TENANT))){
+                SecurityDynamicConfiguration<?> rolesMappingsConfiguration = load(CType.ROLESMAPPING, false);
+                Set<String> userRoles = new HashSet<>();
+                Set<String> accessibleTenants = new HashSet<>();
+
+                // find user roles
+                for (String role : rolesMappingsConfiguration.getCEntries().keySet()){
```

Comment:
There should be an existing functionality which gets user roles

Reply:
Because roles are also present with internalusers and not just in rolesmapping

Reply:
You can use this from the same class instead of again calculating the roles.

`final Set<String> securityRoles = privilegesEvaluator.mapRoles(user, remoteAddress);`